File: baekjoon/2636_치즈/치즈.js
// 골드 4 : https://www.acmicpc.net/problem/2636
const fs = require("fs");
const file = process.platform === "linux" ? "/dev/stdin" : "./text.txt";
const input = fs.readFileSync(file).toString().trim().split("\n");

const inputArr = input.map((value) => value.split(" ").map(Number));

const [n, m] = inputArr.shift();

let graph = [];

for (let i = 0; i < n; i++) {
  graph.push(inputArr[i]);
}

let q = [];
const dx = [-1, 0, 1, 0];
const dy = [0, 1, 0, -1];

const bfs = (row, column, v, check) => {
  q.push([row, column]);
  v[row][column] = true;

  while (q.length) {
    const [x, y] = q.shift();
    let nx = 0;
    let ny = 0;
    for (let i = 0; i < 4; i++) {
      nx = x + dx[i];
      ny = y + dy[i];

      if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
        if (graph[nx][ny] === 0 && !v[nx][ny]) {
          q.push([nx, ny]);
          v[nx][ny] = true;
        }

        if (
          graph[nx][ny] === 1 &&
          !check.some((value) => value[0] === nx && value[1] === ny)
        ) {
          check.push([nx, ny]);
        }
      }
    }
  }
};
const updateMeltGraph = (check) => {
  for (idx of check) {
    const [x, y] = idx;
    graph[x][y] = 0;
  }
};

let result = 0;
let beforeMelt = 0;

while (true) {
  let visited = Array.from({ length: n }, () => Array(m).fill(false));
  let checkIdx = [];

  bfs(0, 0, visited, checkIdx);
  if (checkIdx.length === 0) {
    console.log(result);
    console.log(beforeMelt);
    return;
  }
  if (checkIdx.length > 0) {
    result += 1;
    beforeMelt = checkIdx.length;
  }

  updateMeltGraph(checkIdx);
}

File: baekjoon/2636_치즈/치즈.py
# 골드 4 : https://www.acmicpc.net/problem/2636
from collections import deque

n,m = map(int, input().split())
graph = []
for _ in range(n):
  graph.append(list(map(int, input().split())))

q = deque()
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]


def bfs(row, column):
  q.append((row, column))
  visited[row][column]

  while q:
    x, y = q.popleft()

    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]

      if (nx >= 0 and ny >= 0 and nx < n and ny < m):
        if graph[nx][ny] == 0 and not visited[nx][ny]:
          visited[nx][ny] = True
          q.append((nx, ny))
          
        if graph[nx][ny] == 1 and (nx, ny) not in checkIdx:
          checkIdx.append((nx, ny))


def updateMeltGraph():
  for idx in checkIdx:
    x, y = idx
    graph[x][y] = 0

result = 0
beforeMelt = 0


while True:
  visited = [[False for _ in range(m)] for _ in range(n)]
  checkIdx = []
  bfs(0,0)

  if (len(checkIdx) > 0):
    result += 1
    beforeMelt = len(checkIdx)

  if (len(checkIdx) == 0):
    break

  updateMeltGraph()

print(result)
print(beforeMelt)